flink-connector-aws/flink-connector-aws-kinesis-streams/src/main/java/org/apache/flink/connector/kinesis/source/enumerator/KinesisStreamsSourceEnumerator.java

```diff
@@ -143,17 +156,101 @@ public void addReader(int subtaskId) {
 
     @Override
     public KinesisStreamsSourceEnumeratorState snapshotState(long checkpointId) throws Exception {
-        return new KinesisStreamsSourceEnumeratorState(unassignedSplits, lastSeenShardId);
+        List<KinesisShardSplitWithAssignmentStatus> splitStates =
+                splitTracker.snapshotState(checkpointId);
+        return new KinesisStreamsSourceEnumeratorState(splitStates, lastSeenShardId);
     }
 
     @Override
     public void close() throws IOException {
         streamProxy.close();
     }
 
-    private List<KinesisShardSplit> initialDiscoverSplits() {
-        List<Shard> shards = streamProxy.listShards(streamArn, lastSeenShardId);
-        return mapToSplits(shards, sourceConfig.get(STREAM_INITIAL_POSITION));
+    @VisibleForTesting
```

Comment:
Can we avoid these "VisibleForTesting" methods by modifying the TestKinesisStreamProxy?

Reply:
In this case logic does not fit in the StreamsProxy, but also needs to be tested.
Discussed with @hlteoh37 offline and we agreed that VisibleForTesting is ok here.